### PR TITLE
[Bugfix][Gemma4] Keep image bidirectional attention within the sliding window

### PR DESCRIPTION
**Pull request by lucianommartins**
## Purpose

Gemma4 vision models that use blockwise bidirectional image attention
(`use_bidirectional_attention="vision"` → 12B, 26B-A4B, 31B) silently fell back
to **causal-only** attention over image tokens whenever a single image's
soft-token span exceeded the text sliding window (e.g. ~1100 soft tokens at
`max_soft_tokens=1120` vs `sliding_window=1024`). This degraded vision/OCR
quality — catastrophic on the encoder-free **12B** (no vision tower to
compensate, output collapsed into fragmented `...` text) and latent on the
tower-based 26B/31B.

The defect is backend-agnostic (the runner builds the ranges), so it affected
both TRITON_ATTN (A100/SM80) and FA4 (SM90/Hopper); the fix touches both kernels.

- **Root cause:** `GPUModelRunner` dropped the *entire* bidirectional
  `mm_prefix` range for an image when `(range_len > sliding_window)`, leaving
  those image tokens fully causal in the LM.
- **Intended behavior** (HF reference, https://github.com/huggingface/transformers/pull/46850): on local/sliding
  layers the bidirectional block is **clamped to** the sliding window —
  `(causal OR blockwise) AND sliding_window` — not dropped. Global/full-attention
  layers are causal-only (already handled in vLLM via
  `_clear_mm_prefix_for_full_attn_layers`).

### Fix (three parts, gated to Gemma4)

- **Runner:** gate the oversized-range skip on a new model capability flag
  `mm_prefix_clamp_sliding_window`. Models that bound `mm_prefix` in-kernel keep
  the full range; all other models keep the existing drop (default-unset).
- **Triton kernel (A100/SM80):** add an `MM_PREFIX_CLAMP_SW` `constexpr` to
  `compute_kv_seq_mask`; when set, AND each `mm_prefix` range with the
  sliding-window past-bound `(q - kv) < SLIDING_WINDOW` (equivalent to HF's
  `sliding_window_overlay`; future `kv` passes trivially). Plumbed
  `Attention → TritonAttentionImpl → unified_attention → kernel`. Default
  `False` is dead-code-eliminated, so other models/backends are byte-identical.
- **FA4 / FlashAttention kernel (SM90/Hopper):** Gemma4 is forced onto the
  FLASH_ATTN backend whenever FA4 is available and `max_head_dim ≤ 512`
  (`Gemma4Config.verify_and_update_config`), so the Triton clamp does not run
  there. FA4 builds its mask via a CuTe-DSL `mask_mod` that is the *sole*
  element mask (causal/window masking is off when a `mask_mod` is present), so
  the bidirectional block was not window-bounded. `_make_mm_prefix_mask_mod`
  now takes an optional `sliding_window`; when set it ANDs each range with
  `(q_idx - kv_idx) < sliding_window` — the exact analogue of the Triton clamp.
  `FlashAttentionImpl` passes it only for the same per-layer flag; window =
  `sliding_window_size[0] + 1`.

### Scope / safety

- Gemma4-only, defense-in-depth: the model flag is set only on
  `Gemma4ForConditionalGeneration` (+ its `Gemma4Unified...` subclass); the
  per-layer kernel flag is set only by `Gemma4Attention` on sliding layers
  (`self.is_sliding`).
- The FA4 path is gated by the same per-layer flag: `_make_mm_prefix_mask_mod`
  defaults `sliding_window=0` (clamp dead-code-eliminated at jit-trace time), and
  only Gemma4 sliding layers pass a positive window. Other FA4 mm-prefix models
  (gemma3, etc.) are unchanged.
- **Unaffected:** E2B/E4B (`is_mm_prefix_lm=False`), gemma3 (separate class),
  all non-Gemma models, and any Gemma4 case where the image fits within the
  window or the request is text-only (bit-identical).
- Not a duplicate: no existing open PR addresses the vLLM-side oversized-range
  drop; transformers#46850 is the HF-side counterpart and the reference spec.

## Test Plan

Reference parity against HuggingFace `transformers` (which implements [#46850](https://github.com/huggingface/transformers/pull/46850)),
using a dense OCR image at `max_soft_tokens=1120` (1100 soft tokens > 1024
window), greedy decoding:

- E2E: `gemma-4-12B-it`, `--max-soft-tokens 1120`, temperature 0, original chat
  template; compare generated transcription against the HF reference output.
- Numerical localization (custom harness): dump and diff, at identical inputs,
  (a) vision embeddings, (b) post-merge `inputs_embeds`, (c) first-step logits,
  HF vs vLLM. Ablation: re-run HF with `use_bidirectional_attention` disabled to
  confirm the broken vLLM state == HF-causal.
- SM90/FA4: same E2E + parity check on an H100 box (the harness hooks the model
  forward, not the kernel, so it is backend-agnostic). Verify the log reports the
  FLASH_ATTN/FA4 backend (not FA3/Triton) so the clamped path is exercised.

### A100 / TRITON_ATTN

8×A100-80GB, `TRITON_ATTN` backend (FA4 unavailable on SM80; forced for Gemma4
due to heterogeneous head dims 256/512), CUDA graphs enabled.

- **First-step logits (12B @ 1120), vLLM vs HF reference:**
  - Before: argmax **818**, `MATCH=False` (HF argmax 236825) — i.e. vLLM ==
    HF-with-bidirectional-disabled (confirmed by ablation).
  - After: argmax **236825**, **identical top-5**
    `[236825, 2717, 818, 1018, 236777]`, cosine **0.998**.
- **Vision embeddings** and **post-merge `inputs_embeds`** were already
  bit-identical to HF (`max_abs_diff = 0.0`), isolating the defect to the LM
  attention mask.
- **E2E OCR (12B @ 1120):** before → fragmented `...`-delimited output; after →
  fluent transcription matching the HF reference.
- **Regressions:** 12B @ 280 and text-only are bit-identical before/after.

### H100 / FA4

H100, FlashAttention v4 — log confirms the path is exercised:
`Using FA4 for all layers`, `Using FLASH_ATTN attention backend`,
`Using FlashAttention version 4`.

- **E2E OCR (12B @ 1120), temperature 0:** before → fragmented `...` output;
  after → fluent transcription verbatim-matching the HF reference (same text as
  the A100/Triton run). Confirms the gated CuTe `mm_prefix` sliding-window clamp
  on the FA4 path.

### CODE REVIEW

File: vllm/v1/attention/ops/triton_attention_helpers.py [hunk 1]
```diff
@@ -349,7 +357,10 @@ def compute_kv_seq_mask(
                 & (seq_offset[None, :] <= range_end)
                 & is_valid
             )
-            seq_mask |= q_in_range & k_in_range
+            mm_mask = q_in_range & k_in_range
+            if MM_PREFIX_CLAMP_SW and SLIDING_WINDOW > 0:
+                mm_mask = mm_mask & ((query_abs_pos - seq_offset) < SLIDING_WINDOW)
+            seq_mask |= mm_mask
```
Comment (Isotr0py) [inline vllm/v1/attention/ops/triton_attention_helpers.py:365]: I think this is equivalent to apply mm_prefix mask only for causal attention layer?
```diff
--- if USE_MM_PREFIX:
+++ if USE_MM_PREFIX and not (MM_PREFIX_CLAMP_SW and SLIDING_WINDOW > 0)
        for i in range(MAX_MM_RANGES): ...
```

File: vllm/v1/attention/backends/flash_attn.py [hunk 2]
```diff
@@ -876,7 +876,22 @@ def forward(
                     and self.vllm_flash_attn_version == 4
                 ):
                     max_ranges = mm_prefix_ranges.shape[1]
-                    mm_mask_mod = _make_mm_prefix_mask_mod(max_ranges)
+                    # Gemma4: clamp the bidirectional block to the sliding
+                    # window (HF (causal OR blockwise) AND sliding_window on
+                    # local layers). Other mm-prefix models pass sliding_window=0
+                    # -> unclamped, behavior unchanged. sliding_window_size[0]+1
+                    # is the window (self.sliding_window = (sw-1, 0) on sliding
+                    # layers); causal mm_prefix never hits the symmetric path.
```
Comment (Isotr0py) [inline vllm/v1/attention/backends/flash_attn.py:900]: I think we can simply skip the mm_prefix mask computation for SWA in this case?

File: vllm/v1/attention/ops/triton_attention_helpers.py (hunk 1 above)
Comment (lucianommartins) [reply vllm/v1/attention/ops/triton_attention_helpers.py:365]: thanks for reviewing, @Isotr0py! these aren't equivalent: that guard skips the `mm_prefix`
loop on gemma4's sliding layers and reverts them to causal-only (the bug this
fixes). Two points:

1. The clamp isn't a no-op on sliding layers: the base mask is already
`causal AND sliding_window`... the loop ORs in `blockwise AND (q-kv<SW)`;
the only part not already covered by the base is future in-block keys (`kv > q`)
which is excluded by causal, required for bidirectional image attention. The clamp keeps
them (`kv > q ⇒ q-kv < 0 < SW`, matching HF's `sliding_window_overlay`, which only
bounds the past), giving `(q-kv<SW) AND (causal OR blockwise)` = HF's
`(causal OR blockwise) AND sliding_window` (transformers#46850)...

Example: image block `[0,1099]`, `q=10`, `kv=500` (same image, future):
base `500<=10`=False; clamped mm_prefix `q_in&k_in`=True, `q-kv=-490<SW`=True →
attends... with the guard the loop is skipped → `q=10` never sees `kv=500`

2. the layer roles are inverted: in Gemma4 the bidirectional block belongs on
the sliding/local layers; full/global layers are causal-only and already have
`mm_prefix` cleared upstream (`_clear_mm_prefix_for_full_attn_layers`)... so gating
the loop off for the clamped (sliding) layers leaves bidirectional applied on no
layer at all

File: vllm/v1/attention/backends/flash_attn.py (hunk 2 above)
Comment (lucianommartins) [reply vllm/v1/attention/backends/flash_attn.py:900]: skipping the `mm_prefix` mask_mod on the SWA (sliding) layers would revert them
to plain causal+window and drop the image bidirectional attention — the bug this
fixes...

FA4-specific reason it can't be skipped: on FA4 the mask_mod is the sole
element mask: when a `mask_mod` is set, `apply_mask` turns off its own
causal/window element masking (`cute/mask.py`), so `window_size` only drives
block-level skipping, not the mask... with no mask_mod, FA falls back to native
`causal + window` = causal-only. The blockwise term (an image token attending to
later tokens of the same image, `kv > q`) has no native FA flag; the mask_mod
is the only way to express it.

window and bidirectional are orthogonal: the window bounds distance (past
side), the mask_mod adds the forward in-block direction... the clamp just makes
them consistent: `(kv<=q) OR (in-block AND q-kv<SW)` = HF's
`(causal OR blockwise) AND sliding_window`.

concrete: sliding layer, image block `[0,1099]`, `q=10`, `kv=500` (same image,
future). causal `500<=10`=False; mask_mod `in_block & (q-kv=-490 < SW)`=True →
attends. Drop the mask_mod and `q=10` never sees `kv=500`.

**Comment by cjackal** [comment]
Hi, while I tested this PR, I found that FA4 backend has accuracy issue for long(i.e. input token count above the chunked prefill size) image inputs. I used a test script similar to the one in #38257 and it just spits a gibberish (like a la la la ...).

It seems triton backend (which is the default backend back when FA4 was not available) does not have this accuracy issue, and FA4 also works okay for long text-only inputs, so I wonder if FA4 branch has something wrong in terms of bidirectional masking?

**Comment by lucianommartins** [comment]
Hi @cjackal - it is what I'm fixing here... both Triton and FA4 backend have bidi for vision issues.

if you are still saying bad results even after this, it can be related to the image soft tokens budget: by default it is an economic value of 280 which seems to low to more complex/detailed images - you can test increasing this budget too.

**Comment by cjackal** [comment]
> Hi @cjackal - it is what I'm fixing here... both Triton and FA4 backend have bidi for vision issues.
> 
> 
> 
> if you are still saying bad results even after this, it can be related to the image soft tokens budget: by default it is an economic value of 280 which seems to low to more complex/detailed images - you can test increasing this budget too.

Thanks for the quick reply, what I observed is that (on this PR branch) my test script passes for triton backend (however slow it is) but generates total gibberish when the attention backend is changed to FA4, so I think there's something fundamentally different between triton and FA4 attention backend on computing multimodal tokens. ~~I'll post in a form of proper github issue with minimal reproducer shortly.~~ Done in #47300, as given there I used the default soft token for both attention backend.

**Review by Isotr0py — APPROVED** [review]

**Comment by lucianommartins** [comment]
can we force merge it, @ywang96 @DarkLight1337? @Isotr0py  already approved it, but the automerge is never triggered due to some CI tests issues (unrelated to this PR).
